File: algorithmic_toolbox/week2_algorithmic_warmup/1_fibonacci_number/fibonacci.go
package main

import (
	"C"
	"fmt"
	"testing"
)

func fibonacciNaive(n int) int {
	if n <= 1 {
		return n
	}
	return fibonacciNaive(n-1) + fibonacciNaive(n-2)
}

func fibonacciFast(n int) int {
	if n <= 1 {
		return n
	}

	fibNums := make([]int, n+1)
	fibNums[0] = 0
	fibNums[1] = 1

	for i := 2; i <= n; i++ {
		fibNums[i] = fibNums[i-1] + fibNums[i-2]
	}

	return fibNums[n]
}

func testSolution(t *testing.T) {
	assertEqual := func(a, b int) {
		if a != b {
			t.Errorf("expected %d but got %d", b, a)
		}
	}

	assertEqual(fibonacciFast(3), 2)
	assertEqual(fibonacciFast(10), 55)
	for n := 0; n < 20; n++ {
		assertEqual(fibonacciFast(n), fibonacciNaive(n))
	}
}

func main() {
	var n int
	fmt.Scan(&n)

	fmt.Println(fibonacciNaive(n))
	fmt.Println(fibonacciFast(n))
	testSolution(nil)

	fmt.Println("OK")
}

File: algorithmic_toolbox/week2_algorithmic_warmup/1_fibonacci_number/fibonacci.cpp
#include <cassert>
#include <iostream>
#include <vector>

int fibonacci_naive(int n) {
  return n <= 1 ? n : fibonacci_naive(n - 1) + fibonacci_naive(n - 2);
}

int fibonacci_fast(int n) {
  if (n <= 1) return n;

  std::vector<long> fibNums(n + 1);
  fibNums[0] = 0;
  fibNums[1] = 1;

  for (int i = 2; i <= n; ++i)
    fibNums[i] = fibNums[i - 1] + fibNums[i - 2];

  return fibNums[n];
}

void test_solution() {
  assert(fibonacci_fast(3) == 2);
  assert(fibonacci_fast(10) == 55);
  for (int n = 0; n < 20; ++n) assert(fibonacci_fast(n) == fibonacci_naive(n));
}

int main() {
  int n = 0;
  std::cin >> n;

  std::cout << fibonacci_naive(n) << std::endl;
  std::cout << fibonacci_fast(n) << std::endl;
  test_solution();

  std::cout << "OK" << std::endl;
  return 0;
}
